File: LeetCode/Hard/3307-Find_The_K-th_Character_In_String_Game_II/3307-FindTheK-thCharacterInStringGameII.cpp
class Solution {
public:
    char kthCharacter(long long k, vector<int>& operations) {
        int t = 0;
        long long current_len = 1;
        k--;

        while (current_len <= k) {
            current_len *= 2;
            t++;
        }

        char final_char = 'a';

        while (current_len > 1) {
            current_len /= 2;
            t--;

            if (k >= current_len) {
                k -= current_len;

                if (operations[t] == 1)
                    final_char++;

                if (final_char > 'z')
                    final_char = 'a';
            }
        }

        return final_char;
    }
};

File: LeetCode/Hard/3307-Find_The_K-th_Character_In_String_Game_II/3307-FindTheK-thCharacterInStringGameII.py
class Solution:
    def kthCharacter(self, k: int, operations: list[int]) -> str:
        k -= 1
        t = 0

        current_len = 1
        while current_len <= k:
            current_len *= 2
            t += 1

        final_char = ord("a")

        while current_len > 1:
            current_len //= 2
            t -= 1

            if k >= current_len:
                k -= current_len
                if operations[t] == 1:
                    final_char += 1

                    if final_char > ord("z"):
                        final_char = 97

        return chr(final_char)
